aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -823,14 +821,8 @@ impl VMExecutor for AptosVM {
             ))
         });
 
-        // Execute transactions in parallel if on chain config is set and loaded.
-        if let Some(_read_write_set_analysis) =
-            ParallelExecutionConfig::fetch_config(&RemoteStorage::new(state_view))
-                .and_then(|config| config.read_write_analysis_result)
-                .map(|config| config.into_inner())
-        {
-            // Note that writeset transactions will be executed sequentially as it won't be inferred
-            // by the read write set analysis and thus fall into the sequential path.
+        // Execute transactions in parallel if concurrency level is set to > 1.
+        if envmnt::get_u16("APTOS_EXECUTION_CONCURRENCY_LEVEL", 1) > 1 {
```

Comment:
are we planning on keeping this around post launch? if so, we should definitely put this into the config file: https://github.com/aptos-labs/aptos-core/blob/main/config/src/config/execution_config.rs

Reply:
I'm curious why we hit this if statement each time rather than using some form of static dispatch generated in the constructor of AptosVM?

Reply:
We shouldn't be making the decision each time. This isn't dynamic. I'll need to look at the code but it seems like we can do better here.

Reply:
Some sort of local config seems like the cleanest solution (on chain makes no sense since everyone should decide for themselves?) - but seems like a lot of refactor that we could do later (or now, happy to tackle that if we so decide). A related question is actually why is everything static - we could maybe save some allocations and deallocations later if we had an instance of executor instead of this stateless design, also useful for configurations and such. So let's discuss this, N1, whether we should do a config file now (or later) vs env var (for now).

Off to the env var solution, what is in this PR now is trash. I can make it be less trash by having one function inside executor. But N2 is whether we want dynamic behavior - I thought maybe it's useful to reconfiguring sequential or parallel (or how many threads) on the fly, but I suppose env variables aren't usually used that way and it looks more like a side effect of an already started process to change behavior. So, if we all agree, I'll make it a one time decision (and will use static dispatching on it, will just initialize an enum once lazily).

Reply:
I don't think we need dynamic behavior on the fly, restart to take effect seems reasonable.

what I suggested with config + lazy static is we can parse the config value inside the constructor https://github.com/aptos-labs/aptos-core/blob/b913372f5943683cfa063ac3d65efc4a0a4187c3/consensus/src/consensus_provider.rs#L52 and initialize a lazy static variable later read by the parallel executor.

Reply:
plan:
have a once_cell in Aptos VM that determines concurrency level, we initialize where Zekun pointed out but also set default (that would make it sequential) if get access returns None. It may not be fully static dispatch (per the real meaning of that term), due to once cell implementation, but it will not be a dynamic behavior anymore, the provided execution config inside node config will determine concurrency level once.

Parallel executor constructor will then take concurrency level as parameter, assert > 1, <= num_cpu, and use that. Will change all the tests & benchmarks affected by this change as well, shouldn't be too bad.

Reply:
@davidiw @zekun000 PR updated, please have another look